src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs

```diff
@@ -1205,12 +1206,17 @@ private BoundNode RewriteCatch(BoundCatchBlock node, ArrayBuilder<BoundExpressio
             else if (prologue.Count > 0)
             {
                 Debug.Assert(rewrittenFilter != null);
-                rewrittenFilter = new BoundSequence(
-                    rewrittenFilter.Syntax,
-                    ImmutableArray.Create<LocalSymbol>(),
-                    prologue.ToImmutable(),
-                    rewrittenFilter,
-                    rewrittenFilter.Type);
+                var prologueBuilder = ArrayBuilder<BoundStatement>.GetInstance(prologue.Count);
+                if (rewrittenFilterPrologue != null)
+                {
+                    prologueBuilder.AddRange(rewrittenFilterPrologue.Statements);
+                }
+                foreach (var p in prologue)
+                {
+                    prologueBuilder.Add(new BoundExpressionStatement(p.Syntax, p) { WasCompilerGenerated = true });
+                }
```

Comment:
I'm trying to convince myself that this is the right order between prologues, but I'm not sure.

Reply:
This preserves the order previously in the compiler, it just stores them elsewhere.

A way to see that this order is correct is to recall the purposes of the two kinds of prologues. The one coming in as `rewrittenFilterPrologue` used to be "attached" to the expression so is part of evaluating it, which is why we put it right before the expression. The new prologue sets up a frame for any variables that are captured. Since those variables might be used in the expression and the filter prologue, that has to go first.

Having said that, what I just said doesn't correspond to what the code does. It appears to swap the order (a bug). I'll fix it and add a test showing why the order is important.

---
In reply to: [456718553](https://github.com/dotnet/roslyn/pull/46054#discussion_r456718553) [](ancestors = 456718553)